src/btf.cpp

```diff
@@ -0,0 +1,257 @@
+#include <iostream>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/utsname.h>
+#include <string.h>
+#include <linux/limits.h>
+#include "btf.h"
+#include "types.h"
+#include "bpftrace.h"
+
+#ifdef HAVE_LIBBPF_BTF_DUMP
+#include <linux/bpf.h>
+#include <linux/btf.h>
+#include <bpf/btf.h>
+#include <bpf/libbpf.h>
+
+namespace bpftrace {
+
+static unsigned char *get_data(const char *file, ssize_t *sizep)
+{
+  struct stat st;
+
+  if (stat(file, &st))
+    return NULL;
+
+  FILE *f;
+
+  f = fopen(file, "rb");
+  if (!f)
+    return NULL;
+
+  unsigned char *data;
+  unsigned int size;
+
+  size = st.st_size;
+
+  data = (unsigned char *) malloc(size);
+  if (!data)
+  {
+    fclose(f);
+    return NULL;
+  }
+
+  ssize_t ret = fread(data, 1, st.st_size, f);
+
+  if (ret != st.st_size)
+  {
+    free(data);
+    fclose(f);
+    return NULL;
+  }
+
+  fclose(f);
+
+  *sizep = size;
+  return data;
+}
+
+static struct btf* btf_raw(char *file)
+{
+  unsigned char *data;
+  ssize_t size;
+  struct btf *btf;
+
+  data = get_data(file, &size);
+  if (!data)
+  {
+    std::cerr << "BTF: failed to read data from: " << file << std::endl;
+    return NULL;
+  }
+
+  btf = btf__new(data, (__u32) size);
+  free(data);
+  return btf;
+}
+
+static int libbpf_print(enum libbpf_print_level level, const char *msg, va_list ap)
+{
+  fprintf(stderr, "BTF: (%d) ", level);
+  return vfprintf(stderr, msg, ap);
+}
+
+struct btf_location {
+  const char *path;
+  bool        raw;
+};
+
+static struct btf *btf_open(struct btf_location *locs)
+{
+  struct utsname buf;
+
+  uname(&buf);
+
+  for (int i = 0; locs[i].path; i++)
+  {
+    char path[PATH_MAX + 1];
+
+    snprintf(path, PATH_MAX, locs[i].path, buf.release);
+    if (access(path, R_OK))
+      continue;
+
+    struct btf *btf;
+
+    if (locs[i].raw)
+      btf = btf_raw(path);
+    else
+      btf = btf__parse_elf(path, NULL);
+
+    int err = libbpf_get_error(btf);
+
+    if (err)
+    {
+      if (bt_verbose)
+      {
+        char err_buf[256];
+
+        libbpf_strerror(libbpf_get_error(btf), err_buf, sizeof(err_buf));
+        std::cerr << "BTF: failed to read data (" << err_buf << ") from: " << path << std::endl;
+      }
+      continue;
+    }
+
+    if (bt_verbose)
+    {
+      std::cerr << "BTF: using data from " << path << std::endl;
+    }
+    return btf;
+  }
+
+  return NULL;
+}
+
+BTF::BTF(void) : btf(NULL), state(NODATA)
+{
+  // 'borrowed' from libbpf's bpf_core_find_kernel_btf
+  // from Andrii Nakryiko
+  struct btf_location locs_normal[] =
```

Comment:
Maybe use `std::array` instead. It knows its own size so you can do a rangeed for loop over it

Reply:
ok, will check

Reply:
on the other hand I need to know size of the std::array when
defining it, so it seems more convenient to me to use C array
in here.. however if you want to stick with c++ objects I understand
and can change it, let me know

Reply:
Yeah that's kinda annoying. I think they're trying to add support for it in newer c++: https://en.cppreference.com/w/cpp/experimental/make_array . We can use that then